adapters/src/main/kotlin/com/krzykrucz/elesson/currentlesson/topic/adapters/rest/TopicRouterConfig.kt

```diff
@@ -0,0 +1,71 @@
+package com.krzykrucz.elesson.currentlesson.topic.adapters.rest
+
+
+import arrow.core.Option
+import arrow.core.extensions.option.traverse.sequence
+import arrow.core.fix
+import arrow.core.getOrElse
+import arrow.core.toOption
+import arrow.fx.IO
+import arrow.fx.extensions.fx
+import arrow.fx.extensions.io.applicative.applicative
+import arrow.fx.fix
+import arrow.fx.typeclasses.Duration
+import com.krzykrucz.elesson.currentlesson.attendance.infrastructure.fetchAttendanceIO
+import com.krzykrucz.elesson.currentlesson.domain.attendance.CheckedAttendance
+import com.krzykrucz.elesson.currentlesson.domain.topic.domain.chooseTopic
+import com.krzykrucz.elesson.currentlesson.topic.adapters.persistence.fetchFinishedLessonsCount
+import com.krzykrucz.elesson.currentlesson.topic.adapters.persistence.persistInProgressLesson
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+import org.springframework.http.MediaType
+import org.springframework.web.reactive.function.BodyInserters
+import org.springframework.web.reactive.function.server.ServerRequest
+import org.springframework.web.reactive.function.server.ServerResponse
+import org.springframework.web.reactive.function.server.router
+import reactor.core.publisher.Mono
+import java.util.concurrent.TimeUnit
+
+@Configuration
+class TopicRouterConfig {
+    @Bean
+    fun topicRouter() = router {
+        (path("/topic") and accept(MediaType.APPLICATION_JSON)).nest {
+            POST("", handleChooseTopicRequest())
+        }
+    }
+
+    private fun handleChooseTopicRequest(): (ServerRequest) -> Mono<ServerResponse> = { request ->
+        request
+                .bodyToMono(ChooseTopicDto::class.java)
+                .flatMap { handleChooseTopicDto(it) }
+    }
+
+    private fun handleChooseTopicDto(dto: ChooseTopicDto): Mono<ServerResponse> =
+            IO.fx {
+                val (checkedAttendance) = fetchAttendanceIO()(dto.lessonIdentifier)
+                val (finishedLessonsCount) = fetchFinishedLessonsCount()()
+                val (inProgressLessonOpt) = checkedAttendance
```

Comment:
I'm not sure if we plan to follow a clean hexagonal architecture but this piece of orchestration code (fetching and creating something) sounds for me more like a responsibility of an `application/use case` layer which then can be used by multiple output ports and it's adaptors [like here](https://trello-attachments.s3.amazonaws.com/5dac91cf48b5577f74875176/993x647/3455430002150a72ca2b09dd48635ad8/image.png)

To but it simpler, I think this orchestration should be put somewhere else and then re-used in this REST adaptor.

Reply:
In connection to functional DDD style, I'm not sure how this can be achieved, looks like a missing workflow to do all of this things or something in between to translate DTO from an adaptor to then use predefined workflow :man_shrugging:

Reply:
I've extracted this orchestration code to `use case` package. However, I had problems with persistence implementation, where should it be placed?

Reply:
persistence sounds like a port in `use case` layer which can be implemented in `adapter` layer

Reply:
I would not create a use case package/layer. I would keep everything in either domain or an adapter. I don't want to overcomplicate things.

Reply:
Personally I would keep it the adapter layer